wurst/lib/StackNSplit.wurst

```diff
@@ -0,0 +1,242 @@
+    int slotFrom = getInventorySlotFrom()
+    item with = getInventorySwappedItem()
+    int curr
+    int diff
+    int charges = GetItemCharges(itm)
+    int itemTypeId = itm.getTypeId()
+    int size = u.inventorySize()
+    int max
+    int toSplit
+
+    if not isItemStackable(itemTypeId)
+        return
+
+    max = getItemMaxStacks(itemTypeId)
+    toSplit = getItemSplitCount(itemTypeId)
+
+    if slotFrom == getInventorySlotTo() // splitting
+        if charges < 2 // need at least 2
+            return
+        else if toSplit >= charges
+            toSplit = charges - 1
+        SetItemCharges(itm, charges - toSplit)
+        fireEvent(eventRemoveTrigger, u, itm, toSplit)
+
+        // Redistribute the splitted stacks if possible
+        for i = 0 to size - 1
+            if i != slotFrom
+                with = u.itemInSlot(i)
+                if with.getTypeId() == itemTypeId
+                    curr = GetItemCharges(with)
+                    diff = max - curr
+                    if diff > toSplit
+                        SetItemCharges(with, curr + toSplit)
+                        fireEvent(eventAddedTrigger, u, with, toSplit)
+                    else
+                        SetItemCharges(with, max)
+                        fireEvent(eventAddedTrigger, u, with, diff)
+                    toSplit -= diff
+            if toSplit <= 0
+                break
+        if toSplit > 0 // something is left
+            with = CreateItem(itemTypeId, u.getX(), u.getY())
+            SetItemCharges(with, toSplit)
+            trigger t = getPlayerUnitEventTrigger(EVENT_PLAYER_UNIT_PICKUP_ITEM)
+            t.disable()
+            UnitAddItem(u, with)
+            t.enable()
+    else if with.getTypeId() == itemTypeId
+        curr = GetItemCharges(with)
+        toSplit = charges + curr
+
+        if toSplit > max
+            if curr != max
+                diff = max - charges
+                SetItemCharges(itm, max)
+                fireEvent(eventAddedTrigger, u, itm, diff)
+                SetItemCharges(with, toSplit - max)
+                fireEvent(eventRemoveTrigger, u, with, diff)
+        else
+            SetItemCharges(with, toSplit)
+            fireEvent(eventAddedTrigger, u, with, charges)
+            itm.remove()
+
+function onSmoothPickup()
+    stackItem(getSmoothItemPickupUnit(), getSmoothItemPickupItem())
+
+class StackSmoothPickupPredicate implements SmoothPickupPredicate
+
+    function canPickup(unit u, item itm) returns boolean
+        int itemTypeId = itm.getTypeId()
+        if not isItemStackable(itemTypeId)
+            return false
+
+        return u.isInventoryFull() and not u.isItemFullyStacked(itemTypeId)
+
+init
+    registerPlayerUnitEvent(EVENT_PLAYER_UNIT_PICKUP_ITEM, () -> onPickUp())
+    registerInventoryEvent(EVENT_INVENTORY_MOVED, () -> onMoved())
+    registerSmoothItemPickupEvent(() -> onSmoothPickup())
+    addSmoothPickupCondition(new StackSmoothPickupPredicate())
```

Comment:
We've made some customizations to stack and split which may need to be factored into this version some how. Specifically we made changes to how this library handles diseased meat.

https://github.com/island-troll-tribes/island-troll-tribes/blob/master/jass/systems/EasyItemStacknSplit.j#L139
https://github.com/island-troll-tribes/island-troll-tribes/blob/master/jass/systems/EasyItemStacknSplit.j#L484

Not sure if we have any other changes to the file (other than the stack table rather than item level)